src/main/scala/mesosphere/marathon/api/akkahttp/Directives.scala

```diff
@@ -14,14 +18,60 @@ import scala.concurrent.duration._
   */
 object Directives extends AuthDirectives with LeaderDirectives with AkkaDirectives {
 
+  private val marathonApiKeywords = Set("restart", "tasks", "versions")
+
+  //basically it's equivalent to path.takeWhile(s => !marathonApiKeywords.contains(s)) map createPathId
+  private[akkahttp] def findAppId(remainingPath: Path, extractedAppName: Path = Path.Empty): Option[PathId] = {
+    //transforming Path to PathId
+    def generatePathId(p: Path): PathId = {
+      val trimmed = p.toString().reverse.dropWhile(_ == '/').reverse
+      PathId(trimmed)
+    }
+
+    remainingPath match {
+      case Path.Slash(rest) =>
+        findAppId(rest, extractedAppName ++ Path./)
+      case p @ Path.Segment(segment, rest) =>
+        if (marathonApiKeywords.contains(segment)) {
+          Some(generatePathId(extractedAppName))
+        } else {
+          findAppId(rest, extractedAppName + segment)
+        }
+      case Path.Empty =>
+        if (extractedAppName.isEmpty) {
+          None
+        } else {
+          Some(generatePathId(extractedAppName))
+        }
+    }
+  }
+
   /**
-    * Path matcher, that matches a segment only, if it is defined in the given set.
-    * @param set the allowed path segments.
+    * Given the current root group, only extract an existing appId
```

Comment:
Ok, so let's say we have an app called `restart` in a group called `my-group` and we want to restart this app. We would call `POST /v2/apps/my-group/restart/restart`. What is `extractExistingAppId` returning?

What if I have a group called `restart`?

Reply:
It will extract it correctly. I added a test for this case.